### PR TITLE
refac: Publish events for wholesale results

### PR DESCRIPTION
**Pull request by JesperGomes**

### CODE REVIEW

File: source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/AmountPerChargeResultProducedV1FactoryTests.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+using FluentAssertions;
+using Google.Protobuf.WellKnownTypes;
+using NodaTime;
+using Test.Core;
+using Xunit;
+using QuantityQuality = Energinet.DataHub.Wholesale.CalculationResults.Interfaces.CalculationResults.Model.QuantityQuality;
+
+namespace Energinet.DataHub.Wholesale.Events.UnitTests.Infrastructure.IntegrationEvents.Factories;
+
+public class AmountPerChargeResultProducedV1FactoryTests
```
Comment (dstenroejl) [inline source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/AmountPerChargeResultProducedV1FactoryTests.cs:30]: Maybe we should have a test documenting which properties are important when we compare this message/event with the "Monthly" message/event. By that I mean `ChargeResolution`.

File: source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1FactoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        // Act
+        var act = () => sut.Create(wholesaleResult);
+
+        // Act and Assert
+        act.Should().Throw<ArgumentOutOfRangeException>();
+    }
+
+    [Theory]
+    [InlineAutoMoqData]
+    public void Create_WhenWholesaleResultHsMoreThanOneTimeSeriesPoints_ThrowsException(MonthlyAmountPerChargeResultProducedV1Factory sut)
```
Comment (dstenroejl) [inline source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1FactoryTests.cs:76]:
```suggestion
    public void Create_WhenWholesaleResultHasMoreThanOneTimeSeriesPoints_ThrowsException(MonthlyAmountPerChargeResultProducedV1Factory sut)
```

File: source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1FactoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+using Energinet.DataHub.Wholesale.Events.Infrastructure.IntegrationEvents.Types;
+using FluentAssertions;
+using NodaTime;
+using Test.Core;
+using Xunit;
+using QuantityQuality = Energinet.DataHub.Wholesale.CalculationResults.Interfaces.CalculationResults.Model.QuantityQuality;
+
+namespace Energinet.DataHub.Wholesale.Events.UnitTests.Infrastructure.IntegrationEvents.Factories;
+
+public class MonthlyAmountPerChargeResultProducedV1FactoryTests
```
Comment (dstenroejl) [inline source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1FactoryTests.cs:29]: Maybe we should have a test documenting which properties are important when we compare this message/event with the "Amount" message/event. By that I mean `ChargeResolution`.

**Review by dstenroejl — COMMENTED** [review]
Commented

File: source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1FactoryTests.cs (hunk 3 above)
Comment (JesperGomes) [reply source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1FactoryTests.cs:29]: I have an upcoming PR where I introduce a new propery called AmountType, which is maps directly to the event type. But yes, a test could probably still make sense to have for documentation

File: source/dotnet/wholesale-api/WebApi/Configuration/Modules/EventsRegistration.cs [hunk 4]
```diff
@@ -63,6 +63,10 @@ private static void AddApplications(this IServiceCollection services)
                 CalculationResultCompletedFactory>();
         services.AddScoped<IEnergyResultProducedV1Factory,
             EnergyResultProducedV1Factory>();
+        services.AddScoped<IAmountPerChargeResultProducedV1Factory,
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/WebApi/Configuration/Modules/EventsRegistration.cs:66]: Fixed

File: source/dotnet/wholesale-api/Events/Events.Application/Communication/ICalculationResultIntegrationEventFactory.cs [hunk 5]
```diff
@@ -22,4 +23,8 @@ public interface ICalculationResultIntegrationEventFactory
     IntegrationEvent CreateCalculationResultCompleted(EnergyResult energyResult);
 
     IntegrationEvent CreateEnergyResultProducedV1(EnergyResult energyResult);
+
+    IntegrationEvent CreateAmountPerChargeResultProducedV1(WholesaleResult wholesaleResult);
+
+    IntegrationEvent CreateMonthlyAmountPerChargeResultProducedV1(WholesaleResult wholesaleResult);
 }
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Application/Communication/ICalculationResultIntegrationEventFactory.cs:30]: Won't fix: Potential future improvement

File: source/dotnet/wholesale-api/Events/Events.Application/Communication/IntegrationEventProvider.cs [hunk 6]
```diff
@@ -57,19 +62,52 @@ public async IAsyncEnumerable<IntegrationEvent> GetAsync()
...
+        {
+            ChargeResolution.Day or ChargeResolution.Hour => _calculationResultIntegrationEventFactory
+                .CreateAmountPerChargeResultProducedV1(wholesaleResult),
+            ChargeResolution.Month => _calculationResultIntegrationEventFactory
+                .CreateMonthlyAmountPerChargeResultProducedV1(wholesaleResult),
+            _ => throw new ArgumentOutOfRangeException(nameof(wholesaleResult.ChargeResolution), actualValue: wholesaleResult.ChargeResolution, "Unexpected resolution."),
+        };
+    }
+
+    private static bool IsWholesaleCalculationType(ProcessType calculationType)
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Application/Communication/IntegrationEventProvider.cs:112]: Won't fix: Potential future improvement

File: source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/AmountPerChargeResultProducedV1Factory.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+using Energinet.DataHub.Wholesale.Contracts.IntegrationEvents;
+using Energinet.DataHub.Wholesale.Events.Infrastructure.IntegrationEvents.Mappers.AmountPerChargeResultProducedV1;
+using Energinet.DataHub.Wholesale.Events.Infrastructure.IntegrationEvents.Types;
+using Google.Protobuf.WellKnownTypes;
+
+namespace Energinet.DataHub.Wholesale.Events.Infrastructure.IntegrationEvents.Factories;
+
+public class AmountPerChargeResultProducedV1Factory : IAmountPerChargeResultProducedV1Factory
+{
+    public AmountPerChargeResultProducedV1 Create(WholesaleResult result)
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/AmountPerChargeResultProducedV1Factory.cs:25]: Won't fix: Potential future improvement

File: source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1Factory.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Energinet.DataHub.Wholesale.Events.Infrastructure.IntegrationEvents.Factories;
+
+public class MonthlyAmountPerChargeResultProducedV1Factory : IMonthlyAmountPerChargeResultProducedV1Factory
+{
+    public MonthlyAmountPerChargeResultProducedV1 Create(WholesaleResult result)
+    {
+        if (result.TimeSeriesPoints.Count != 1)
+            throw new ArgumentException("MonthlyAmountPerChargeResultProducedV1 expects exactly one time series point.");
+
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/MonthlyAmountPerChargeResultProducedV1Factory.cs:28]: Won't fix: Potential future improvement

File: source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/CalculationResultIntegrationEventFactory.cs [hunk 9]
```diff
@@ -46,6 +55,20 @@ public IntegrationEvent CreateEnergyResultProducedV1(EnergyResult energyResult)
             return CreateIntegrationEvent(calculationResultCompleted, eventIdentification, EnergyResultProducedV1.EventName, EnergyResultProducedV1.EventMinorVersion);
         }
 
+        public IntegrationEvent CreateAmountPerChargeResultProducedV1(WholesaleResult wholesaleResult)
+        {
+            var calculationResultCompleted = _amountPerChargeResultProducedV1Factory.Create(wholesaleResult);
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/CalculationResultIntegrationEventFactory.cs:60]: Won't fix

File: source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/CalculationResultIntegrationEventFactory.cs [hunk 10]
```diff
@@ -46,6 +55,20 @@ public IntegrationEvent CreateEnergyResultProducedV1(EnergyResult energyResult)
...
+            var eventIdentification = Guid.NewGuid();
+            return CreateIntegrationEvent(calculationResultCompleted, eventIdentification, AmountPerChargeResultProducedV1.EventName, AmountPerChargeResultProducedV1.EventMinorVersion);
+        }
+
+        public IntegrationEvent CreateMonthlyAmountPerChargeResultProducedV1(WholesaleResult wholesaleResult)
+        {
+            var calculationResultCompleted = _monthlyAmountPerChargeResultProducedV1Factory.Create(wholesaleResult);
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/Factories/CalculationResultIntegrationEventFactory.cs:67]: Won't fix

File: source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/MonthlyAmountPerChargeResultProducedV1.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ReSharper disable once CheckNamespace - the namespace is exposed publicly in the Contracts package
+namespace Energinet.DataHub.Wholesale.Contracts.IntegrationEvents;
+
+public partial class MonthlyAmountPerChargeResultProducedV1
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/MonthlyAmountPerChargeResultProducedV1.cs:18]: Won't fix: Potential future improvement

File: source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/AmountPerChargeResultProducedV1.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ReSharper disable once CheckNamespace - the namespace is exposed publicly in the Contracts package
+namespace Energinet.DataHub.Wholesale.Contracts.IntegrationEvents;
+
+public partial class AmountPerChargeResultProducedV1
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.Infrastructure/IntegrationEvents/AmountPerChargeResultProducedV1.cs:18]: Won't fix: Potential future improvement

File: source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/AmountPerChargeResultProducedV1FactoryTests.cs (hunk 1 above)
Comment (dstenroejl) [reply source/dotnet/wholesale-api/Events/Events.UnitTests/Infrastructure/IntegrationEvents/Factories/AmountPerChargeResultProducedV1FactoryTests.cs:30]: Won't fix: Potential future improvement

**Review by dstenroejl — APPROVED** [review]
LGTM

**Pull request merged**